### PR TITLE
Add a nnwf to vmc++

### PR DESCRIPTION
**Pull request by francesco086**
As from the title, this branch introduces the NN support for the Wave Functions.
Trello card:

Since wave functions using a NN will be at the heart of our work, it makes sense to build a Wave Function class that uses a Feed Forward Neural Network. In this way we will not need to re-write it for every example.

AC:
- a documented class must be inside the library
- there must be a running example showing how it works
- possibly think of some unittest

### CODE REVIEW

File: config_template.sh [hunk 1]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+OS_NAME=$(uname)
```
Comment (Ithanil) [inline config_template.sh:3]: Mhh I prefer keeping the OS_NAME in the build.sh, since it is not manually set anyway. Keep user settings in config.

Comment (Ithanil) [reply config_template.sh:3]: But not really important.

File: src/FFNNWaveFunction.hpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class FFNNWaveFunction: public WaveFunction{
+
+private:
+    FeedForwardNeuralNetwork * _ffnn;
+
+public:
+    // --- Constructor and destructor
+    // IMPORTANT: The provided ffnn should be ready to use (connected) and have the first, second and variational derivatives substrates
```
Comment (Ithanil) [inline src/FFNNWaveFunction.hpp:18]: I would add another constructor, where you only pass number of layer and units and the ffnn gets built by the constructor. But that's just a feature request.

Comment (Ithanil) [reply src/FFNNWaveFunction.hpp:18]: So if you want a default style nn quickly, you use the second constructor. For custom built NN, pass it.

File: src/FFNNWaveFunction.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+double FFNNWaveFunction::d1(const int &i, const double *in){
+    _ffnn->setInput(_ffnn->getNInput(), in);
+    _ffnn->FFPropagate();
+    return _ffnn->getFirstDerivative(1, i)/_ffnn->getOutput(1);
+}
+
+
+double FFNNWaveFunction::d2(const int &i, const int &j, const double *in){
+    if (i != j)
```
Comment (Ithanil) [inline src/FFNNWaveFunction.cpp:54]: I think we should change the abstract parent class and remove the j, and allow only pure second derivs. As you said: When are we going to need it?
Of course maybe we still want to keep the possibility to use them in the future...

File: debug/unittest/ut1/main.cpp [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+         Compute:    d/dalpha_i log(Psi(x))
+         where alpha are the variational parameters, in our case an array of dimension 1: alpha = (b)
+         */
+         if (i==0){
+            return (2.*_b*(x[0]-_a));
+         } else if (i==1){
+            return (-(x[0]-_a)*(x[0]-_a));
+         } else{
+            using namespace std;
+            cout << "ERRORE vd1 QuadrExponential! " << endl;
```
Comment (Ithanil) [inline debug/unittest/ut1/main.cpp]: Just no. You know what I mean :P

File: debug/unittest/ut1/main.cpp [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+
+
+
+    // create the NN with the right structure and parameters
+    FeedForwardNeuralNetwork * ffnn = new FeedForwardNeuralNetwork(2, 2, 2);
+    GaussianActivationFunction * gss_actf = new GaussianActivationFunction();
+    IdentityActivationFunction * id_actf = new IdentityActivationFunction();
+    ffnn->setLayerActivationFunction(1, gss_actf);
+    ffnn->setLayerActivationFunction(2, id_actf);
```
Comment (Ithanil) [inline debug/unittest/ut1/main.cpp:173]: Just curious: Why and when should I use these? Why did you do it like that here?

**Review by Ithanil — APPROVED** [review]
Just tiny things in the comments, can be merged.

File: config_template.sh (hunk 1 above)
Comment (francesco086) [reply config_template.sh:3]: Pro of having it in config:
- all the other building scripts have this variable automatically
Contra:
- indeed the config file might be intended only as a place for the variables that the user has to set... but not necessarily, in my opinion

In the end I think it's better having it in the config file, for laziness reasons!

File: debug/unittest/ut1/main.cpp (hunk 4 above)
Comment (francesco086) [reply debug/unittest/ut1/main.cpp]: I am not sure actually :) could you tell me?

File: debug/unittest/ut1/main.cpp (hunk 5 above)
Comment (francesco086) [reply debug/unittest/ut1/main.cpp:173]: It is explained at the beginning of the file, in the comment. I did that for having a NN that actually reproduces a gaussian

File: src/FFNNWaveFunction.cpp (hunk 3 above)
Comment (francesco086) [reply src/FFNNWaveFunction.cpp:54]: we should remove it yes, but that's another Trello story (already on the board btw)

File: src/FFNNWaveFunction.hpp (hunk 2 above)
Comment (francesco086) [reply src/FFNNWaveFunction.hpp:18]: mmm... no, I wouldn't do it here. If that's the case, then we should add it to the FFNN library.
You can add this point to the "Make FFNN more user friendly" Trello's card

File: config_template.sh (hunk 1 above)
Comment (Ithanil) [reply config_template.sh:3]: Good point.

File: debug/unittest/ut1/main.cpp (hunk 5 above)
Comment (Ithanil) [reply debug/unittest/ut1/main.cpp:173]: Might be useful for my fitting?!

**Pull request merged**